sfge/src/main/java/com/salesforce/rules/UnusedMethodRule.java

```diff
@@ -1,155 +1,159 @@
-import org.apache.tinkerpop.gremlin.structure.Vertex;
 
-public class UnusedMethodRule extends AbstractStaticRule {
+/**
+ * Rule for identifying dead code. After all {@link ApexPath} instances have been traversed,
+ * generates violations for any eligible {@link MethodVertex} that was never invoked. A method is
+ * considered eligible if it meets all the following criteria:
+ *
+ * <ol>
+ *   <li>The method is not declared as a {@code testMethod}
+ *   <li>The method is not declared as {@code abstract}
+ *   <li>The method is not a private, 0-arity constructor
+ *   <li>The method is not a VisualForce getter/setter
+ *   <li>The method is not considered a path entrypoint
+ *   <li>The method is not annotated with an engine directive that silences this rule.
+ * </ol>
+ */
+public final class UnusedMethodRule extends AbstractPathTraversalRule
+        implements PostProcessingRule {
+    private static final Logger LOGGER = LogManager.getLogger(UnusedMethodRule.class);
     private static final String URL =
             "https://forcedotcom.github.io/sfdx-scanner/en/v3.x/salesforce-graph-engine/rules/#UnusedMethodRule";
-    private static final Logger LOGGER = LogManager.getLogger(UnusedMethodRule.class);
-    private static final String DESCRIPTION = "Identifies methods that are not invoked";
-    private static final String VIOLATION_TEMPLATE = "Method %s in class %s is never invoked";
-
-    GraphTraversalSource g;
-    /** A helper object used to track state and caching as the rule executes. */
-    RuleStateTracker ruleStateTracker;
+    /**
+     * Used internally for tracking every unique method invoked during the traversal of every path.
+     */
+    private Set<String> encounteredUsageKeys;
 
     private UnusedMethodRule() {
-        super();
-    }
-
-    public static UnusedMethodRule getInstance() {
-        return LazyHolder.INSTANCE;
-    }
-
-    @Override
-    protected int getSeverity() {
-        return SEVERITY.LOW.code;
-    }
-
-    @Override
-    protected String getDescription() {
-        return DESCRIPTION;
+        this.encounteredUsageKeys = CollectionUtil.newTreeSet();
     }
 
     @Override
-    protected String getCategory() {
-        return CATEGORY.PERFORMANCE.name;
+    public boolean test(BaseSFVertex vertex) {
+        // This rule is interested in any vertex representing an invocation of a method.
+        return vertex instanceof InvocableVertex;
     }
 
-    @Override
-    protected String getUrl() {
-        return URL;
-    }
-
-    @Override
-    protected boolean isEnabled() {
-        return true;
-    }
-
-    public RuleStateTracker getRuleStateTracker() {
-        return ruleStateTracker;
+    /**
+     * Reset the internal state of the rule, so it can be reused freshly. NOTE: This method should
+     * only be invoked in tests. TODO: Can we make this method's existence unnecessary?
+     */
+    @VisibleForTesting
+    public void reset() {
```

Comment:
Since the rule has to maintain state, we need a way to reset it between tests. In a perfect world, this would be unnecessary, but I can't figure out how we'd do it.

Reply:
Agree, ideally would be cleaner to not have to maintain state.

Reply:
Might be safer to keep the `encounteredUsageKeys` in its own singleton class that's protected by a provider. Have a look at `ProgressListener`, `ProgressListenerImpl`, and `ProgressListenerProvider`. The interface in this case can have a method to add keys so that all of the addition happens in a thread-safe way.